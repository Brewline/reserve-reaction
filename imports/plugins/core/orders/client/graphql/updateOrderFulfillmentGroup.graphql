updateOrderFulfillmentGroup($orderFulfillmentGroupId: ID!, $orderId: ID!, $status: String, $tracking: String) {
  updateOrderFulfillmentGroup(input: {
    orderFulfillmentGroupId: $orderFulfillmentGroupId,
    orderId: $orderId,
    status: $status,
    tracking: $tracking
  }) {
    order {
      fulfillmentGroups {
        status
        tracking
      }
    }
  }
}
