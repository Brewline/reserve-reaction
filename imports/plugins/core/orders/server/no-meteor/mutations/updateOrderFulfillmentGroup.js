import SimpleSchema from "simpl-schema";
import ReactionError from "@reactioncommerce/reaction-error";
import { Order as OrderSchema } from "/imports/collections/schemas";

const inputSchema = new SimpleSchema({
  tracking: {
    type: String,
    optional: true
  },
  trackingUrl: {
    type: String,
    optional: true
  },
  orderFulfillmentGroupId: String,
  orderId: String,
  status: {
    type: String,
    optional: true
  }
});

/**
 * @method updateOrderFulfillmentGroup
 * @summary Use this mutation to update an order fulfillment group status and tracking information
 * @param {Object} context - an object containing the per-request state
 * @param {Object} input - Necessary input. See SimpleSchema
 * @return {Promise<Object>} Object with `order` property containing the updated order
 */
export default async function updateOrderFulfillmentGroup(context, input) {
  inputSchema.validate(input);

  const {
    tracking,
    trackingUrl,
    orderFulfillmentGroupId,
    orderId,
    status
  } = input;

  const { appEvents, collections, isInternalCall, userHasPermission, userId } = context;
  const { Orders } = collections;

  // First verify that this order actually exists
  const order = await Orders.findOne({ _id: orderId });
  if (!order) throw new ReactionError("not-found", "Order not found");

  // Allow update if the account has "orders" permission. When called internally by another
  // plugin, context.isInternalCall can be set to `true` to disable this check.
  if (!isInternalCall && !userHasPermission(["orders"], order.shopId)) {
    throw new ReactionError("access-denied", "Access Denied");
  }

  // Verify that there is a group with the ID
  const orderFulfillmentGroup = order.shipping.find((group) => group._id === orderFulfillmentGroupId);
  if (!orderFulfillmentGroup) throw new ReactionError("not-found", "Order fulfillment group not found");

  const modifier = {
    $set: {
      "shipping.$.updatedAt": new Date(),
      "updatedAt": new Date()
    }
  };

  if (tracking) modifier.$set["shipping.$.tracking"] = tracking;
  if (trackingUrl) modifier.$set["shipping.$.trackingUrl"] = trackingUrl;

  if (status && orderFulfillmentGroup.workflow.status !== status) {
    modifier.$set["shipping.$.workflow.status"] = status;
    modifier.$push = {
      "shipping.$.workflow.workflow": status
    };
  }

  // Skip updating if we have no updates to make
  if (Object.keys(modifier.$set).length === 2) return { order };

  OrderSchema.validate(modifier, { modifier: true });

  const { modifiedCount, value: updatedOrder } = await Orders.findOneAndUpdate(
    {
      "_id": orderId,
      "shipping._id": orderFulfillmentGroupId
    },
    modifier,
    { returnOriginal: false }
  );
  if (modifiedCount === 0 || !updatedOrder) throw new ReactionError("server-error", "Unable to update order");

  await appEvents.emit("afterOrderUpdate", {
    order: updatedOrder,
    updatedBy: userId
  });

  return { order: updatedOrder };
}
