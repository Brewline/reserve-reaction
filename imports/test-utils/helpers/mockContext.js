const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  shopId: "FAKE_SHOP_ID",
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "Groups",
  "Inventory",
  "MediaRecords",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "Tags",
  "Templates",
  "Themes",
  "Translations",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
