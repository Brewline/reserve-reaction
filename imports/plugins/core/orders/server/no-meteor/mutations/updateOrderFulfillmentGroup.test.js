import updateOrderFulfillmentGroup from "./updateOrderFulfillmentGroup";
import Factory from "/imports/test-utils/helpers/factory";
import mockContext from "/imports/test-utils/helpers/mockContext";

beforeEach(() => {
  jest.resetAllMocks();
});

test("throws if orderId isn't supplied", async () => {
  await expect(updateOrderFulfillmentGroup(mockContext, { orderFulfillmentGroupId: "123" })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if orderFulfillmentGroupId isn't supplied", async () => {
  await expect(updateOrderFulfillmentGroup(mockContext, { orderId: "123" })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if the order doesn't exist", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve(null));

  await expect(updateOrderFulfillmentGroup(mockContext, {
    orderId: "order1",
    orderFulfillmentGroupId: "123"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if the order fulfillment group doesn't exist", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      {
        _id: "abc",
        items: []
      }
    ],
    shopId: "SHOP_ID"
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  await expect(updateOrderFulfillmentGroup(mockContext, {
    orderId: "order1",
    orderFulfillmentGroupId: "123"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if permission check fails", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      {
        _id: "123",
        items: []
      }
    ],
    shopId: "SHOP_ID"
  }));

  mockContext.userHasPermission.mockReturnValueOnce(false);

  await expect(updateOrderFulfillmentGroup(mockContext, {
    orderId: "order1",
    orderFulfillmentGroupId: "123"
  })).rejects.toThrowErrorMatchingSnapshot();

  expect(mockContext.userHasPermission).toHaveBeenCalledWith(["orders"], "SHOP_ID");
});

test("skips permission check if context.isInternalCall", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      Factory.OrderFulfillmentGroup.makeOne({
        _id: "123",
        items: Factory.OrderItem.makeMany(2)
      }),
      Factory.OrderFulfillmentGroup.makeOne({
        items: Factory.OrderItem.makeMany(2)
      })
    ],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.collections.Orders.findOneAndUpdate.mockReturnValueOnce(Promise.resolve({
    modifiedCount: 1,
    value: {}
  }));

  mockContext.isInternalCall = true;

  await updateOrderFulfillmentGroup(mockContext, {
    orderId: "order1",
    orderFulfillmentGroupId: "123"
  });

  delete mockContext.isInternalCall;

  expect(mockContext.userHasPermission).not.toHaveBeenCalled();
});

test("skips update if one is not necessary", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      Factory.OrderFulfillmentGroup.makeOne({
        _id: "group1",
        items: Factory.OrderItem.makeMany(2)
      })
    ],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  await updateOrderFulfillmentGroup(mockContext, { orderId: "order1", orderFulfillmentGroupId: "group1" });

  expect(mockContext.collections.Orders.findOneAndUpdate).not.toHaveBeenCalled();
});

test("updates an order fulfillment group", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      Factory.OrderFulfillmentGroup.makeOne({
        _id: "group1",
        items: Factory.OrderItem.makeMany(2),
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      })
    ],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  mockContext.collections.Orders.findOneAndUpdate.mockReturnValueOnce(Promise.resolve({
    modifiedCount: 1,
    value: {}
  }));

  await updateOrderFulfillmentGroup(mockContext, {
    orderId: "order1",
    orderFulfillmentGroupId: "group1",
    status: "NEW_STATUS",
    tracking: "TRACK_REF",
    trackingUrl: "http://track.me/TRACK_REF"
  });

  expect(mockContext.collections.Orders.findOneAndUpdate).toHaveBeenCalledWith(
    {
      "_id": "order1",
      "shipping._id": "group1"
    },
    {
      $set: {
        "shipping.$.tracking": "TRACK_REF",
        "shipping.$.trackingUrl": "http://track.me/TRACK_REF",
        "shipping.$.updatedAt": jasmine.any(Date),
        "shipping.$.workflow.status": "NEW_STATUS",
        "updatedAt": jasmine.any(Date)
      },
      $push: {
        "shipping.$.workflow.workflow": "NEW_STATUS"
      }
    },
    { returnOriginal: false }
  );
});
