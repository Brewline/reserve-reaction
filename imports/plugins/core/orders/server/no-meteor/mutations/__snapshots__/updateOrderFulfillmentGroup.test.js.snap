// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if orderFulfillmentGroupId isn't supplied 1`] = `"Order fulfillment group ID is required"`;

exports[`throws if orderId isn't supplied 1`] = `"Order ID is required"`;

exports[`throws if permission check fails 1`] = `"Access Denied"`;

exports[`throws if the order doesn't exist 1`] = `"Order not found"`;

exports[`throws if the order fulfillment group doesn't exist 1`] = `"Order fulfillment group not found"`;
